### PR TITLE
[Inductor] add missing ops for cpp vectorization overrides

### PR DESCRIPTION
**Pull request by Valentine233**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #90750

For micro-benchmark, aten.elu.default and aten.elu_backward.default have poor performance with inductor compared to eager. The main reason is lack of the vectorization. With adding missing ops for cpp vectorization overrides, the vectorization could be successfully applied.

Performance data for eager v.s. inductor:
<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:///C:/Users/xuanliao/AppData/Local/Temp/msohtmlclip1/01/clip.htm">
<link rel=File-List
href="file:///C:/Users/xuanliao/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">

</head>

<body link="#0563C1" vlink="#954F72">

op | speedup_old | RSD (3) | speedup_new | RSD (3) | increased_performance
-- | -- | -- | -- | -- | --
aten.elu.default | 0.205947276 | 1.73% | 0.995302802 | 4.76% | 383.28%
aten.elu_backward.default | 0.336280639 | 0.58% | 1.69473642 | 1.96% | 403.96%

</body>

</html>

The new supported ops for cpp vectorization overrides:
- eq
- ne
- lt
- gt
- le
- ge

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @Guobing-Chen @chunyuan-w @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

### CODE REVIEW

**Review by jgong5 — APPROVED** [review]
LGTM. Is there UT covering these ops?

File: test/inductor/test_torchinductor.py [hunk 1]
```diff
@@ -5244,6 +5244,36 @@ def fn(x):
                 assert same(fn(x)[0], compiled([x])[0], equal_nan=True)
                 assert metrics.generated_cpp_vec_kernel_count == 1
 
+        @unittest.skipIf(
+            not codecache.valid_vec_isa_list(), "Does not support vectorization"
+        )
+        @patch("torch.cuda.is_available", lambda: False)
+        def test_vec_comapre_op_cpu_only(self):
```
Comment (jgong5) [inline test/inductor/test_torchinductor.py:5251]: nit: 
```suggestion
        def test_vec_compare_op_cpu_only(self):
```

File: test/inductor/test_torchinductor.py [hunk 2]
```diff
@@ -5244,6 +5244,36 @@ def fn(x):
...
+
+            x = torch.randn((2, 9))
+
+            with patch.object(config.cpp, "simdlen", None):
+                torch._dynamo.reset()
+                metrics.reset()
+                traced = make_fx(fn)(x)
+                compiled = compile_fx_inner(traced, [x])
+                assert same(fn(x)[0], compiled([x])[0], equal_nan=True)
+                assert metrics.generated_cpp_vec_kernel_count == 1
```
Comment (jgong5) [inline test/inductor/test_torchinductor.py:5275]: Do we also need to assert that non vec kernel count is 0 here?

**Review by jgong5 — CHANGES_REQUESTED** [review]

File: test/inductor/test_torchinductor.py (hunk 2 above)
Comment (Valentine233) [reply test/inductor/test_torchinductor.py:5275]: Sure, is it fine to assert that generated_kernel_count is also 1, so that there is no non vec kernel?

Comment (Valentine233) [reply test/inductor/test_torchinductor.py:5275]: Another assert has been added: (generated_kernel_count - generated_cpp_vec_kernel_count) is 0

File: torch/_inductor/codegen/cpp.py [hunk 3]
```diff
@@ -231,6 +231,30 @@ def erf(x):
...
+        return f"{x} > {y}"
+
+    @staticmethod
+    def le(x, y):
+        return f"{x} <= {y}"
+
+    @staticmethod
+    def ge(x, y):
+        return f"{x} >= {y}"
+
```
Comment (EikanWang) [inline torch/_inductor/codegen/cpp.py:257]: Have you verified TorchBench, HF and TIMM? Meanwhile, I would prefer conservative optimization from the functionality perspective. Could you please share which model requires this optimization?

File: test/inductor/test_torchinductor.py [hunk 4]
```diff
@@ -5244,6 +5244,40 @@ def fn(x):
...
                 assert metrics.generated_cpp_vec_kernel_count == 1
 
+        @unittest.skipIf(
+            not codecache.valid_vec_isa_list(), "Does not support vectorization"
+        )
+        @patch("torch.cuda.is_available", lambda: False)
+        def test_vec_compare_op_cpu_only(self):
+            def fn(x):
+                y1 = torch.eq(x, 1)
+                x = torch.where(y1, x, -x)
```
Comment (EikanWang) [inline test/inductor/test_torchinductor.py:5254]: May I know the data type of `1`?

File: test/inductor/test_torchinductor.py [hunk 5]
```diff
@@ -5244,6 +5244,40 @@ def fn(x):
                 assert same(fn(x)[0], compiled([x])[0], equal_nan=True)
                 assert metrics.generated_cpp_vec_kernel_count == 1
 
+        @unittest.skipIf(
+            not codecache.valid_vec_isa_list(), "Does not support vectorization"
+        )
+        @patch("torch.cuda.is_available", lambda: False)
+        def test_vec_compare_op_cpu_only(self):
+            def fn(x):
+                y1 = torch.eq(x, 1)
```
Comment (EikanWang) [inline test/inductor/test_torchinductor.py:5253]: Please also add UT: `x == 1`. And the same philosophy is for `ne`, `le`, `ge`, `gt`, `lt`

**Review by EikanWang — CHANGES_REQUESTED** [review]

**Review by jgong5 — APPROVED** [review]
LGTM.

File: torch/_inductor/codegen/cpp.py (hunk 3 above)
Comment (Valentine233) [reply torch/_inductor/codegen/cpp.py:257]: `aten.elu.default` needs `gt` and `aten.elu_backward.default` needs `le` for vectorization. The models containing these ops are `nvidia_deeprecommender` in TorchBench and `dpn107` in TIMM.

File: test/inductor/test_torchinductor.py (hunk 4 above)
Comment (Valentine233) [reply test/inductor/test_torchinductor.py:5254]: The data type of `1` is unclear. I would explicitly define `1` as `float32` by replacing `1` with `1.0`. The same for the other constants.

**Review by EikanWang — APPROVED** [review]

File: torch/_inductor/codegen/cpp.py (hunk 3 above)
Comment (Valentine233) [reply torch/_inductor/codegen/cpp.py:257]: This PR has been verified for all models of TorchBench, HF and TIMM.

**Review by jansel — APPROVED** [review]

File: torch/_inductor/codegen/cpp.py (hunk 3 above)
Comment (desertfire) [reply torch/_inductor/codegen/cpp.py:257]: Just to be clear, what does "verified" mean here? Do you mean performance or correctness?

**Review by desertfire — APPROVED** [review]

File: torch/_inductor/codegen/cpp.py (hunk 3 above)
Comment (Valentine233) [reply torch/_inductor/codegen/cpp.py:257]: "Verified" means correctness here.

**Pull request closed without merging**